sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -370,6 +371,45 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       )
     }
 
+    df.sparkSession.sessionState.sqlParser.parseMultipartIdentifier(tableName) match {
+      case CatalogObjectIdentifier(Some(catalog), ident) =>
+        insertInto(catalog, ident)
+      case AsTableIdentifier(tableIdentifier) =>
+        insertInto(tableIdentifier)
+    }
+  }
+
+  private def insertInto(catalog: CatalogPlugin, ident: Identifier): Unit = {
+    import org.apache.spark.sql.catalog.v2.CatalogV2Implicits._
+
+    val table = DataSourceV2Relation.create(catalog.asTableCatalog.loadTable(ident))
+
+    val command = modeForDSV2 match {
+      case SaveMode.Append =>
+        AppendData.byName(table, df.logicalPlan)
```

Comment:
I missed it. If you look at the doc of `insertInto`, it says
```
   * Inserts the content of the `DataFrame` to the specified table. It requires that
   * the schema of the `DataFrame` is the same as the schema of the table.
   *
   * @note Unlike `saveAsTable`, `insertInto` ignores the column names and just uses position-based
   * resolution. For example:
```

We should use `byPosition` here.

Reply:
Agreed. This is an oversight and should be by position.

Reply:
Thanks @cloud-fan @rdblue. I will submit a hotfix.

Reply:
I can create a follow-up PR to introduce an option `matchByName`, default to `false`. If true, insertInto uses byName; otherwise, byPosition.

Maybe even included in this PR?

Reply:
If we are going to create a new dataframe writer API in the future, I'd like to keep it as it is, and always do by-position in this `insertInto`.

Reply:
> If we are going to create a new dataframe writer API in the future, I'd like to keep it as it is, and always do by-position in this insertInto.

Sounds good to me. I'll submit a PR for the new API.

Reply:
@rdblue @cloud-fan @dongjoon-hyun #25353 is ready for review.